File: 3.Longest Common Substring/LCS.py
# Python implementation of Longest Common Substring

def LCS_Iterative(s1, s2):
    res = 0
    n, m = len(s1), len(s2)
    for i in range(n):
        for j in range(m):
            curr = 0
            while i + curr < n and j + curr < m and s1[i + curr] == s2[j + curr]:
                curr += 1
            res = max(res, curr)
    return res

def LCS_Solve(s1, s2, n, m):
    if n == 0 or m == 0 or s1[n - 1] != s2[m - 1]:
        return 0
    return 1 + LCS_Solve(s1, s2, n - 1, m - 1)

def LCS_Recursion(s1, s2):
    res = 0
    for i in range(1, len(s1) + 1):
        for j in range(1, len(s2) + 1):
            res = max(res, LCS_Solve(s1, s2, i, j))
    return res

def LCS_Tabulation(s1, s2):
    res = 0
    m, n = len(s1), len(s2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
                res = max(res, dp[i][j])
            else:
                dp[i][j] = 0
    return res

def LCS_Tabulation_SpaceOptimized(s1, s2):
    res = 0
    n, m = len(s1), len(s2)
    prev = [0] * (m + 1)
    cur = [0] * (m + 1)

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                cur[j] = prev[j - 1] + 1
                res = max(res, cur[j])
            else:
                cur[j] = 0
        prev = cur[:]
    return res

def LCS_Tabulation_Print(s1, s2):
    m, n = len(s1), len(s2)
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    max_len = 0
    end_index = -1

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if s1[i - 1] == s2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]
                if dp[i][j] > max_len:
                    max_len = dp[i][j]
                    end_index = i - 1
            else:
                dp[i][j] = 0

    lcs = ""
    if max_len > 0:
        lcs = s1[end_index - max_len + 1:end_index + 1]

    print(f"Longest Common Substring: {lcs} || and length is: {max_len}")

    return lcs

def LCS_Tabulation_SpaceOptimized_Print(s1, s2):
    n, m = len(s1), len(s2)
    prev = [0] * (m + 1)
    cur = [0] * (m + 1)
    max_len = 0
    end_index = -1

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if s1[i - 1] == s2[j - 1]:
                cur[j] = prev[j - 1] + 1
                if cur[j] > max_len:
                    max_len = cur[j]
                    end_index = i - 1
            else:
                cur[j] = 0
        prev = cur[:]

    lcs = ""
    if max_len > 0:
        lcs = s1[end_index - max_len + 1:end_index + 1]

    print(f"Longest Common Substring: {lcs} || and length is: {max_len}")

    return lcs

# Main function to test the implementation
def main():
    s1 = "geeksforgeeks"
    s2 = "practicewritegeekscourses"

    print("Longest Common Substring Iterative:", LCS_Iterative(s1, s2))
    print("Longest Common Substring Recursion:", LCS_Recursion(s1, s2))
    print("Longest Common Substring Tabulation:", LCS_Tabulation(s1, s2))
    print("Longest Common Substring Tabulation_SpaceOptimized:", LCS_Tabulation_SpaceOptimized(s1, s2))

    LCS_Tabulation_Print(s1, s2)
    LCS_Tabulation_SpaceOptimized_Print(s1, s2)

if __name__ == "__main__":
    main()

File: 3.Longest Common Substring/LCS.js
// JavaScript implementation of Longest Common Substring

function LCS_Iterative(s1, s2) {
  let res = 0;
  let n = s1.length,
    m = s2.length;
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < m; j++) {
      let curr = 0;
      while (i + curr < n && j + curr < m && s1[i + curr] === s2[j + curr])
        curr++;
      res = Math.max(res, curr);
    }
  }
  return res;
}

function LCS_Solve(s1, s2, n, m) {
  if (n === 0 || m === 0 || s1[n - 1] !== s2[m - 1]) return 0;
  return 1 + LCS_Solve(s1, s2, n - 1, m - 1);
}

function LCS_Recursion(s1, s2) {
  let res = 0;
  for (let i = 1; i <= s1.length; i++)
    for (let j = 1; j <= s2.length; j++)
      res = Math.max(res, LCS_Solve(s1, s2, i, j));
  return res;
}

function LCS_Tabulation(s1, s2) {
  let res = 0;
  let m = s1.length,
    n = s2.length;
  let dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        dp[i][j] = 1 + dp[i - 1][j - 1];
        res = Math.max(res, dp[i][j]);
      } else {
        dp[i][j] = 0;
      }
    }
  }
  return res;
}

function LCS_Tabulation_SpaceOptimized(s1, s2) {
  let res = 0;
  let n = s1.length,
    m = s2.length;
  let prev = Array(m + 1).fill(0);
  let cur = Array(m + 1).fill(0);

  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        cur[j] = prev[j - 1] + 1;
        res = Math.max(res, cur[j]);
      } else {
        cur[j] = 0;
      }
    }
    prev = [...cur];
  }
  return res;
}

function LCS_Tabulation_Print(s1, s2) {
  let m = s1.length,
    n = s2.length;
  let dp = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));
  let maxLen = 0,
    endIndex = -1;

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        dp[i][j] = 1 + dp[i - 1][j - 1];
        if (dp[i][j] > maxLen) {
          maxLen = dp[i][j];
          endIndex = i - 1;
        }
      } else {
        dp[i][j] = 0;
      }
    }
  }

  let lcs = "";
  if (maxLen > 0) lcs = s1.substring(endIndex - maxLen + 1, endIndex + 1);

  console.log(
    "Longest Common Substring: " + lcs + " || and length is: " + maxLen
  );

  return lcs;
}

function LCS_Tabulation_SpaceOptimized_Print(s1, s2) {
  let n = s1.length,
    m = s2.length;
  let prev = Array(m + 1).fill(0);
  let cur = Array(m + 1).fill(0);
  let maxLen = 0,
    endIndex = -1;

  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      if (s1[i - 1] === s2[j - 1]) {
        cur[j] = prev[j - 1] + 1;
        if (cur[j] > maxLen) {
          maxLen = cur[j];
          endIndex = i - 1;
        }
      } else {
        cur[j] = 0;
      }
    }
    prev = [...cur];
  }

  let lcs = "";
  if (maxLen > 0) lcs = s1.substring(endIndex - maxLen + 1, endIndex + 1);

  console.log(
    "Longest Common Substring: " + lcs + " || and length is: " + maxLen
  );

  return lcs;
}

// Main function to test the implementation
(function main() {
  let s1 = "geeksforgeeks";
  let s2 = "practicewritegeekscourses";

  console.log("Longest Common Substring Iterative: " + LCS_Iterative(s1, s2));
  console.log("Longest Common Substring Recursion: " + LCS_Recursion(s1, s2));
  console.log("Longest Common Substring Tabulation: " + LCS_Tabulation(s1, s2));
  console.log(
    "Longest Common Substring Tabulation_SpaceOptimized: " +
      LCS_Tabulation_SpaceOptimized(s1, s2)
  );

  LCS_Tabulation_Print(s1, s2);
  LCS_Tabulation_SpaceOptimized_Print(s1, s2);
})();
